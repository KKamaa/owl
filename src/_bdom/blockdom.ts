import {
  Block,
  BLOCK_TYPE,
  Builder,
  Anchor,
  BlockElement,
  BlockMulti,
  BlockList,
  BlockText,
} from "./types";

// -----------------------------------------------------------------------------
//  Factories
// -----------------------------------------------------------------------------

export function text(text: string): BlockText {
  return {
    type: BLOCK_TYPE.Text,
    el: undefined,
    key: undefined,
    data: text,
    content: undefined,
  };
}

export function elem(builder: Builder, data: any[] = [], children: Block[] = []): BlockElement {
  return {
    type: BLOCK_TYPE.Element,
    el: undefined,
    key: undefined,
    data: { builder, data },
    content: children,
  };
}

export function multi(blocks: (Block | undefined)[]): BlockMulti {
  return {
    type: BLOCK_TYPE.Multi,
    el: undefined,
    key: undefined,
    data: undefined,
    content: blocks,
  };
}

export function list(
  blocks: Block[],
  isOnlyChild: boolean = false,
  hasNoComponent: boolean = false
): BlockList {
  return {
    type: BLOCK_TYPE.List,
    el: undefined,
    key: undefined,
    data: { anchor: undefined, isOnlyChild, hasNoComponent },
    content: blocks,
  };
}
// -----------------------------------------------------------------------------
//  BDom manipulation
// -----------------------------------------------------------------------------

export function mountBefore(block: Block, anchor: Anchor) {
  switch (block.type) {
    case BLOCK_TYPE.Text:
      {
        let el = document.createTextNode(block.data);
        block.el = el;
        anchor.before(el);
      }
      break;
    case BLOCK_TYPE.Element:
      {
        const data = block.data;
        const builder = new data.builder(data, block.content);
        data.builder = builder;
        const el = builder.el;
        anchor.before(el);
        block.el = el;
      }
      break;
    case BLOCK_TYPE.Multi:
      {
        const children = block.content;
        const anchors = new Array(children.length);
        for (let i = 0, l = children.length; i < l; i++) {
          let child = children[i];
          const childAnchor = document.createTextNode("");
          anchor.before(childAnchor);
          anchors![i] = childAnchor;
          if (child) {
            mountBefore(child, childAnchor);
          }
        }
        block.data = anchors;
      }
      break;
    case BLOCK_TYPE.List: {
      const children = block.content;
      const _anchor = document.createTextNode("");
      block.data.anchor = _anchor;
      anchor.before(_anchor);
      for (let i = 0, l = children.length; i < l; i++) {
        mountBefore(children[i], _anchor);
      }
    }
  }
}

function moveBefore(block: Block, anchor: Anchor | ChildNode) {
  switch (block.type) {
    case BLOCK_TYPE.Text:
      {
        anchor.before(block.el!);
      }
      break;
    case BLOCK_TYPE.Element:
      {
        anchor.before(block.el!);
      }
      break;
    case BLOCK_TYPE.Multi:
      {
        // console.warn('movebefore', block, (anchor as any).innerHTML)
        const children = block.content;
        const anchors = block.data!;
        for (let i = 0, l = children.length; i < l; i++) {
          let child = children[i];
          let _anchor = anchors[i];
          // console.warn(child!.el!.nextSibling === _anchor)
          // console.warn('c', child)
          anchor.before(_anchor);
          if (child) {
            moveBefore(child, _anchor);
          }
        }
      }
      break;
    case BLOCK_TYPE.List: {
      // const children = block.content;
      // const _anchor = document.createTextNode("");
      // block.data.anchor = _anchor;
      // anchor.before(_anchor);
      // for (let i = 0, l = children.length; i < l; i++) {
      //   mountBefore(children[i], _anchor);
      // }
    }
  }
}

export function removeBlock(block: Block) {
  switch (block.type) {
    case BLOCK_TYPE.Text:
    case BLOCK_TYPE.Element:
      block.el!.remove();
      break;
    case BLOCK_TYPE.Multi:
      {
        const children = block.content;
        const anchors = block.data;
        for (let i = 0; i < children.length; i++) {
          const child = children[i];
          if (child) {
            remove(child);
          }
          anchors![i].remove();
        }
      }
      break;
    case BLOCK_TYPE.List: {
      const { isOnlyChild, anchor } = block.data;
      if (isOnlyChild) {
        anchor!.parentElement!.textContent = "";
      } else {
        const children = block.content;
        for (let i = 0, l = children.length; i < l; i++) {
          removeBlock(children[i]);
        }
        anchor!.remove();
      }
    }
  }
}

export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  switch (block1.type) {
    case BLOCK_TYPE.Text:
      {
        let text = block2.data as any;
        if (block1.data !== text) {
          block1.data = text;
          block1.el!.textContent = text;
        }
      }
      break;
    case BLOCK_TYPE.Element:
      (block1 as any).data.builder.update((block2 as any).data, block2.content);
      break;
    case BLOCK_TYPE.Multi:
      {
        const children = block1.content;
        const newChildren = block2.content!;
        const anchors = block1.data!;
        for (let i = 0, l = children.length; i < l; i++) {
          const block = children[i];
          const newBlock = newChildren[i];
          if (block) {
            if (newBlock) {
              patch(block, newBlock);
            } else {
              children[i] = undefined;
              remove(block);
            }
          } else if (newBlock) {
            children[i] = newBlock;
            mountBefore(newBlock, anchors[i]);
          }
        }
        // block1.content = block2.content!;
      }
      break;
    case BLOCK_TYPE.List:
      {
        const oldCh = block1.content;
        const newCh: Block[] = (block2 as any).content;
        const data = block1.data;
        const _anchor = data.anchor!;

        // fast path
        if (newCh.length === 0 && data.isOnlyChild) {
          if (!data.hasNoComponent) {
            for (let i = 0; i < oldCh.length; i++) {
              beforeRemove(oldCh[i]);
            }
          }

          const parent = _anchor.parentElement!;
          _anchor.remove();
          parent.textContent = "";
          parent.appendChild(_anchor);
          block1.content = newCh;
          return;
        }

        let oldStartIdx = 0;
        let newStartIdx = 0;
        let oldStartBlock = oldCh[0];
        let newStartBlock = newCh[0];

        let oldEndIdx = oldCh.length - 1;
        let newEndIdx = newCh.length - 1;
        let oldEndBlock = oldCh[oldEndIdx];
        let newEndBlock = newCh[newEndIdx];

        let mapping: any = undefined;
        let noFullRemove = data.hasNoComponent;

        // console.warn('oldch', oldCh[0].content)
        while (oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx) {
          // console.warn('iteration', oldStartIdx, oldEndIdx, newStartIdx, newEndIdx);
          // -------------------------------------------------------------------
          if (oldStartBlock === null) {
            // todo: comment this and check it is useful
            oldStartBlock = oldCh[++oldStartIdx];
          }
          // -------------------------------------------------------------------
          else if (oldEndBlock === null) {
            oldEndBlock = oldCh[--oldEndIdx];
          }
          // -------------------------------------------------------------------
          else if (newStartBlock === null) {
            // todo: understand why these 2 next elseifs are useful
            newStartBlock = newCh[++newStartIdx];
          }
          // -------------------------------------------------------------------
          else if (newEndBlock === null) {
            newEndBlock = newCh[--newEndIdx];
          }
          // -------------------------------------------------------------------
          else if (oldStartBlock.key === newStartBlock.key) {
            // console.warn('same start', oldStartBlock, newStartBlock);
            patch(oldStartBlock, newStartBlock);
            newCh[newStartIdx] = oldStartBlock;
            oldStartBlock = oldCh[++oldStartIdx];
            newStartBlock = newCh[++newStartIdx];
          }
          // -------------------------------------------------------------------
          else if (oldEndBlock.key === newEndBlock.key) {
            // console.warn('same end');
            patch(oldEndBlock, newEndBlock);
            // console.warn('oeb', oldEndBlock)
            newCh[newEndIdx] = oldEndBlock;
            // console.warn(newCh[newEndIdx])
            oldEndBlock = oldCh[--oldEndIdx];
            newEndBlock = newCh[--newEndIdx];
          }
          // -------------------------------------------------------------------
          else if (oldStartBlock.key === newEndBlock.key) {
            // bnode moved right
            patch(oldStartBlock, newEndBlock);
            const nextChild = newCh[newEndIdx + 1];
            const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
            moveBefore(oldStartBlock, anchor);
            newCh[newEndIdx] = oldStartBlock;
            oldStartBlock = oldCh[++oldStartIdx];
            newEndBlock = newCh[--newEndIdx];
          }
          // -------------------------------------------------------------------
          else if (oldEndBlock.key === newStartBlock.key) {
            // bnode moved left
            patch(oldEndBlock, newStartBlock);
            const nextChild = oldCh[oldStartIdx];
            const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
            moveBefore(oldEndBlock, anchor);
            newCh[newStartIdx] = oldEndBlock;
            oldEndBlock = oldCh[--oldEndIdx];
            newStartBlock = newCh[++newStartIdx];
          }
          // -------------------------------------------------------------------
          else {
            // console.warn('in else')
            mapping = mapping || createMapping(oldCh, oldStartIdx, oldEndIdx);
            // console.warn('mapping', mapping)
            let idxInOld = mapping[newStartBlock.key];
            if (idxInOld === undefined) {
              // new element
              mountBefore(
                newStartBlock,
                oldStartBlock.el || (firstChildNode(oldStartBlock)! as any)
              );
              // console.log(document.body.innerHTML)
            } else {
              const elmToMove = oldCh[idxInOld];
              moveBefore(elmToMove, oldStartBlock.el || (firstChildNode(oldStartBlock) as any));
              patch(elmToMove, newStartBlock);
              newCh[newStartIdx] = elmToMove;
              oldCh[idxInOld] = null as any;
            }
            newStartBlock = newCh[++newStartIdx];
          }
        }
        // ---------------------------------------------------------------------
        // console.warn('after')
        if (oldStartIdx <= oldEndIdx || newStartIdx <= newEndIdx) {
          if (oldStartIdx > oldEndIdx) {
            const nextChild = newCh[newEndIdx + 1];
            // console.warn('next', nextChild)
            const anchor = nextChild ? nextChild.el || firstChildNode(nextChild)! : _anchor;
            for (let i = newStartIdx; i <= newEndIdx; i++) {
              mountBefore(newCh[i], anchor as any);
            }
          } else {
            for (let i = oldStartIdx; i <= oldEndIdx; i++) {
              let ch = oldCh[i];
              if (ch) {
                if (noFullRemove) {
                  remove(ch);
                } else {
                  removeBlock(ch);
                }
              }
            }
          }
        }
        block1.content = newCh;
      }
      break;
  }
}

function firstChildNode(block: Block): ChildNode | null {
  switch (block.type) {
    case BLOCK_TYPE.Multi:
      {
        const children = block.content;
        for (let i = 0, l = children.length; i < l; i++) {
          const child = children[i];
          if (child) {
            return child.el || firstChildNode(child);
          }
        }
      }
      break;
    case BLOCK_TYPE.List: {
      const first = block.content[0];
      return first ? first.el || firstChildNode(first) : null;
    }
  }
  return null;
}

export function remove(block: Block) {
  beforeRemove(block);
  removeBlock(block);
}

function createMapping(
  oldCh: any[],
  oldStartIdx: number,
  oldEndIdx: number
): { [key: string]: any } {
  let mapping: any = {};
  for (let i = oldStartIdx; i <= oldEndIdx; i++) {
    mapping[oldCh[i].key] = i;
  }
  return mapping;
}
