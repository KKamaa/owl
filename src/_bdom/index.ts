import { Block } from "./types";

export { makeBlock } from "./element";
export { list } from "./list";
export { multi } from "./multi";
export { text } from "./text";

// export function beforeRemove(block: Block) {}

// -----------------------------------------------------------------------------
//  BDom main entry points
// -----------------------------------------------------------------------------

export function mount(block: Block, target: HTMLElement) {
  const anchor = document.createTextNode("");
  target.appendChild(anchor);
  block.ops.mountBefore(block, anchor);
  anchor.remove();
}

export function patch(block1: Block, block2: Block) {
  block1.ops.patch(block1, block2);
}

export function remove(block: Block) {
  // beforeRemove(block);
  block.ops.remove(block);
}
