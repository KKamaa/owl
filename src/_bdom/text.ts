import { Anchor, BlockText, Operations } from "./types";

// -----------------------------------------------------------------------------
//  Text blocks
// -----------------------------------------------------------------------------

export function text(text: string): BlockText {
  return {
    ops: TEXT_OPS,
    el: undefined,
    data: text,
    content: undefined,
  };
}

const TEXT_OPS: Operations = {
  mountBefore(block: any, anchor: Anchor) {
    let el = document.createTextNode(block.data);
    block.el = el;
    anchor.before(el);
  },
  patch(block1: any, block2: any) {
    if (block1 === block2) {
      return;
    }
    let text = block2.data as any;
    if (block1.data !== text) {
      block1.data = text;
      block1.el!.textContent = text;
    }
  },
  moveBefore(block: any, anchor: any) {
    anchor.before(block.el);
  },
  remove(block: any) {
    const el = block.el!;
    el.parentElement!.removeChild(el);
  },
  firstChildNode(block: any): ChildNode | null {
    return block.el;
  },
};
