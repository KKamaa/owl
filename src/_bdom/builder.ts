import { mountBefore, patch, remove } from "./blockdom";
import { Builder } from "./types";

// -----------------------------------------------------------------------------
//  makeBuilder
// -----------------------------------------------------------------------------
interface BuilderContext {
  path: string[];
  isDynamic: boolean;
  hasChild: boolean;
  hasHandler: boolean;
  info: { index: number; type: "text" | "child" | "handler"; path: string; event?: string }[];
}

function toDom(node: ChildNode, ctx: BuilderContext): HTMLElement | Text | Comment {
  switch (node.nodeType) {
    case 1: {
      // HTMLElement
      const tagName = (node as Element).tagName;
      if (tagName.startsWith("owl-text-")) {
        const index = parseInt(tagName.slice(9), 10);
        ctx.info.push({ index, path: ctx.path.join("."), type: "text" });
        ctx.isDynamic = true;
        return document.createTextNode("");
      }
      if (tagName.startsWith("owl-child-")) {
        const index = parseInt(tagName.slice(10), 10);
        ctx.info.push({ index, type: "child", path: ctx.path.join(".") });
        ctx.hasChild = true;
        return document.createTextNode("");
      }
      const result = document.createElement((node as Element).tagName);
      const attrs = (node as Element).attributes;
      for (let i = 0; i < attrs.length; i++) {
        const attrName = attrs[i].name;
        const attrValue = attrs[i].value;
        if (attrName.startsWith("owl-handler-")) {
          const index = parseInt(attrName.slice(12), 10);
          ctx.info.push({ index, path: ctx.path.join("."), type: "handler", event: attrValue });
          ctx.hasHandler = true;
        } else {
          result.setAttribute(attrs[i].name, attrValue);
        }
      }
      let children = (node as Element).childNodes;
      const initialPath = ctx.path.slice();
      let currentPath = initialPath.slice();
      for (let i = 0; i < children.length; i++) {
        currentPath = currentPath.concat(i === 0 ? "firstChild" : "nextSibling");
        ctx.path = currentPath;
        result.appendChild(toDom(children[i], ctx));
      }
      ctx.path = initialPath;
      return result;
    }
    case 3: {
      // text node
      return document.createTextNode(node.textContent!);
    }
    case 8: {
      // comment node
      return document.createComment(node.textContent!);
    }
  }
  throw new Error("boom");
}

export function makeBuilder(str: string): Builder {
  const info: BuilderContext["info"] = [];
  const ctx: BuilderContext = {
    path: ["el"],
    info,
    hasChild: false,
    hasHandler: false,
    isDynamic: false,
  };

  const doc = new DOMParser().parseFromString(str, "text/xml");
  const template = toDom(doc.firstChild!, ctx) as any;
  const isDynamic = ctx.isDynamic;
  const hasChild = ctx.hasChild;
  const hasHandler = ctx.hasHandler;

  // console.log(context)
  const signature = hasChild ? "data, children" : isDynamic || hasHandler ? "data" : "";
  const code: string[] = [
    `  return class Builder {`,
    `    constructor(${signature}) {`,
    `      this.el = template.cloneNode(true);`,
  ];

  // preparing all internal refs
  if (info.length) {
    for (let i = 0; i < info.length; i++) {
      code.push(`      let ref${i} = this.${info[i].path};`);
    }
    code.push(
      `      this.refs = [${info.map((t: any, index: number) => `ref${index}`).join(", ")}];`
    );
  }

  if (isDynamic || hasHandler) {
    code.push(`      this.data = [];`);
  }
  if (hasChild) {
    code.push(`      this.children = [];`);
  }
  if (hasHandler) {
    // todo
    code.push(`      const handler = this.handleEvent;`);
    for (let i = 0; i < info.length; i++) {
      if (info[i].type === "handler") {
        code.push(
          `      ref${i}.addEventListener("${info[i].event}", handler.bind(this, ${info[i].index}));`
        );
      }
    }
  }
  if (info.length) {
    code.push(`      this.update(${signature});`);
  }

  // end of constructor
  code.push(`    }`);
  code.push(`    update(${signature}) {`);

  // update function
  if (info.length) {
    if (isDynamic || hasChild) {
      code.push(`      const refs = this.refs;`);
    }
    if (isDynamic || hasHandler) {
      code.push(`      const currentData = this.data;`);
      code.push(`      const nextData = data.data;`);
    }
    if (hasChild) {
      code.push(`      const currentChildren = this.children;`);
    }

    for (let i = 0; i < info.length; i++) {
      const data = info[i];
      const index = data.index;
      switch (data.type) {
        case "text":
          code.push(
            `      if (nextData[${index}] !== currentData[${index}]) { refs[${i}].textContent = nextData[${index}]}`
          );
          break;
        case "child":
          code.push(
            `      let currentChild${index} = currentChildren[${index}], child${index} = children[${index}];`
          );
          code.push(
            `      if (currentChild${index}) { if (child${index}) { patch(currentChild${index}, child${index}); } else { remove(currentChild${index}); } }`
          );
          code.push(`      else if (child${index}) { mountBefore(child${index}, refs[${i}]) }`);
      }
    }
    if (isDynamic || hasHandler) {
      code.push(`      this.data = nextData;`);
    }
    if (hasChild) {
      code.push(`      this.children = children;`);
    }
  }
  code.push(`    }`);
  if (hasHandler) {
    code.push(`    handleEvent(n) {`);
    code.push(`      const [owner, method] = this.data[n];`);
    code.push(`      owner[method]();`);
    code.push(`    }`);
  }
  code.push(`  }`);

  //   console.warn(code.join("\n"));
  const wrapper = new Function("template, mountBefore, patch, remove", code.join("\n"));
  return wrapper(template, mountBefore, patch, remove);
}
