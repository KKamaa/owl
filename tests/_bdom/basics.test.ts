import { elem, mount, multi, patch, remove, text } from "../../src/_bdom/blockdom";
import { makeBuilder as origMakeBuilder } from "../../src/_bdom/builder";
import { makeTestFixture } from "../helpers";

function makeBuilder(str: string) {
  const B = origMakeBuilder(str);
  expect(B.toString()).toMatchSnapshot();
  return B;
}

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;

beforeEach(() => {
  fixture = makeTestFixture();
});

afterEach(() => {
  fixture.remove();
});

describe("text and elem blocks", () => {
  test("simple text block", async () => {
    const bdom = text("foo");
    expect(bdom.el).toBe(undefined);
    mount(bdom, fixture);
    expect(bdom.el).not.toBe(undefined);
    expect(fixture.innerHTML).toBe("foo");
  });

  test("patching a simple text block", async () => {
    const bdom1 = text("foo");
    const bdom2 = text("bar");
    mount(bdom1, fixture);
    expect(fixture.innerHTML).toBe("foo");
    patch(bdom1, bdom2);
    expect(fixture.innerHTML).toBe("bar");
  });

  test("simple elem block", async () => {
    const builder = makeBuilder("<div>foo</div>");
    const bdom = elem(builder);
    expect(bdom.el).toBe(undefined);

    mount(bdom, fixture);
    expect(bdom.el).not.toBe(undefined);
    expect(fixture.innerHTML).toBe("<div>foo</div>");
  });

  test("a text block can be removed", async () => {
    const bdom = text("cat");
    expect(fixture.childNodes.length).toBe(0);
    mount(bdom, fixture);
    expect(fixture.innerHTML).toBe("cat");
    expect(fixture.childNodes.length).toBe(1);
    remove(bdom);
    expect(fixture.innerHTML).toBe("");
    expect(fixture.childNodes.length).toBe(0);
  });

  test("elem block can be removed", async () => {
    const builder = makeBuilder("<div>foo</div>");
    const bdom = elem(builder);
    mount(bdom, fixture);
    expect(fixture.innerHTML).toBe("<div>foo</div>");
    remove(bdom);
    expect(fixture.innerHTML).toBe("");
    expect(fixture.childNodes.length).toBe(0);
  });

  test("simple block with multiple roots", async () => {
    const builder1 = makeBuilder("<div>foo</div>");
    const builder2 = makeBuilder("<span>bar</span>");

    const bdom = multi([elem(builder1), elem(builder2)]);

    mount(bdom, fixture);
    expect(fixture.innerHTML).toBe("<div>foo</div><span>bar</span>");
  });

  test("block with dynamic content", async () => {
    const builder = makeBuilder("<div><p><owl-text-0/></p></div>");
    const bdom1 = elem(builder, ["foo"]);

    mount(bdom1, fixture);
    expect(fixture.innerHTML).toBe("<div><p>foo</p></div>");
    const bdom2 = elem(builder, ["bar"]);
    patch(bdom1, bdom2);
    expect(fixture.innerHTML).toBe("<div><p>bar</p></div>");
  });

  test("block with 2 dynamic text nodes", async () => {
    const builder = makeBuilder("<div><p><owl-text-0/></p><span><owl-text-1/></span></div>");
    const bdom = elem(builder, ["foo", "bar"]);

    mount(bdom, fixture);
    expect(fixture.innerHTML).toBe("<div><p>foo</p><span>bar</span></div>");
    patch(bdom, elem(builder, ["appa", "yip yip"]));
    expect(fixture.innerHTML).toBe("<div><p>appa</p><span>yip yip</span></div>");
  });

  test("block with subblock", async () => {
    const builder1 = makeBuilder("<div><owl-child-0/></div>");
    const builder2 = makeBuilder("<p>yip yip</p>");
    const bdom = elem(builder1, [], [elem(builder2)]);

    mount(bdom, fixture);

    expect(fixture.innerHTML).toBe("<div><p>yip yip</p></div>");
  });

  test("block with 2 subblocks", async () => {
    const builder1 = makeBuilder("<div><owl-child-0/><owl-child-1/></div>");
    const builder2 = makeBuilder("<p>yip yip</p>");
    const bdom = elem(builder1, [], [elem(builder2), text("appa")]);

    mount(bdom, fixture);

    expect(fixture.innerHTML).toBe("<div><p>yip yip</p>appa</div>");
  });

  test("block with subblock with siblings", async () => {
    const builder1 = makeBuilder("<div><p>1</p><owl-child-0/><p>2</p></div>");
    const builder2 = makeBuilder("<p>yip yip</p>");

    const bdom = elem(builder1, [], [elem(builder2)]);

    mount(bdom, fixture);
    expect(fixture.innerHTML).toBe("<div><p>1</p><p>yip yip</p><p>2</p></div>");
  });

  test("block with conditional child", async () => {
    const builder1 = makeBuilder("<div><p><owl-child-0/></p></div>");
    const builder2 = makeBuilder("<span>foo</span>");

    const tree = elem(builder1);
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p></p></div>");

    patch(tree, elem(builder1, [], [elem(builder2)]));
    expect(fixture.innerHTML).toBe("<div><p><span>foo</span></p></div>");

    patch(tree, elem(builder1));
    expect(fixture.innerHTML).toBe("<div><p></p></div>");
  });

  test("block with subblock with dynamic content", async () => {
    const builder1 = makeBuilder("<div><owl-child-0/></div>");
    const builder2 = makeBuilder("<p><owl-text-0/></p>");

    const tree = elem(builder1, [], [elem(builder2, ["yip yip"])]);

    mount(tree, fixture);

    expect(fixture.innerHTML).toBe("<div><p>yip yip</p></div>");

    patch(tree, elem(builder1, [], [elem(builder2, ["foo"])]));
    expect(fixture.innerHTML).toBe("<div><p>foo</p></div>");
  });

  test("block with dynamic content and subblock", async () => {
    const builder1 = makeBuilder("<div><owl-child-0/><p><owl-text-0/></p></div>");
    const builder2 = makeBuilder("<p>sub block</p>");

    const tree = elem(builder1, ["yip yip"], [elem(builder2)]);
    mount(tree, fixture);
    expect(fixture.innerHTML).toBe("<div><p>sub block</p><p>yip yip</p></div>");

    patch(tree, elem(builder1, ["foo"], [elem(builder2)]));
    expect(fixture.innerHTML).toBe("<div><p>sub block</p><p>foo</p></div>");
  });
});

describe("misc", () => {
  test("reusing a block skips patching process", async () => {
    const foo = text("foo");
    const bar = text("bar");
    let fooCounter = 0;
    let barCounter = 0;
    let fooValue = "foo";
    let barValue = "bar";
    Object.defineProperty(foo, "data", {
      get() {
        fooCounter++;
        return fooValue;
      },
    });
    Object.defineProperty(bar, "data", {
      get() {
        barCounter++;
        return barValue;
      },
      set(val) {
        barValue = val;
      },
    });

    const bdom = multi([foo, bar]);
    mount(bdom, fixture);
    expect(fooCounter).toBe(1);
    expect(barCounter).toBe(1);
    expect(fixture.innerHTML).toBe("foobar");

    patch(bdom, multi([foo, text("otherbar")]));
    expect(fixture.innerHTML).toBe("foootherbar");
    expect(fooCounter).toBe(1);
    expect(barCounter).toBe(2);
  });
});
