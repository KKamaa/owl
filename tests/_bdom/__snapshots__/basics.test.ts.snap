// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`text and elem blocks block with 2 dynamic text nodes 1`] = `
"class Builder {
    constructor(data) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.firstChild;
      let ref1 = this.el.firstChild.nextSibling.firstChild;
      this.refs = [ref0, ref1];
      this.data = [];
      this.update(data);
    }
    update(data) {
      const refs = this.refs;
      const currentData = this.data;
      const nextData = data.data;
      if (nextData[0] !== currentData[0]) { refs[0].textContent = nextData[0]}
      if (nextData[1] !== currentData[1]) { refs[1].textContent = nextData[1]}
      this.data = nextData;
    }
  }"
`;

exports[`text and elem blocks block with 2 subblocks 1`] = `
"class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      let ref1 = this.el.firstChild.nextSibling;
      this.refs = [ref0, ref1];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); } }
      else if (child0) { mountBefore(child0, refs[0]) }
      let currentChild1 = currentChildren[1], child1 = children[1];
      if (currentChild1) { if (child1) { patch(currentChild1, child1); } else { remove(currentChild1); } }
      else if (child1) { mountBefore(child1, refs[1]) }
      this.children = children;
    }
  }"
`;

exports[`text and elem blocks block with 2 subblocks 2`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks block with conditional child 1`] = `
"class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.firstChild;
      this.refs = [ref0];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); } }
      else if (child0) { mountBefore(child0, refs[0]) }
      this.children = children;
    }
  }"
`;

exports[`text and elem blocks block with conditional child 2`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks block with dynamic content 1`] = `
"class Builder {
    constructor(data) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.firstChild;
      this.refs = [ref0];
      this.data = [];
      this.update(data);
    }
    update(data) {
      const refs = this.refs;
      const currentData = this.data;
      const nextData = data.data;
      if (nextData[0] !== currentData[0]) { refs[0].textContent = nextData[0]}
      this.data = nextData;
    }
  }"
`;

exports[`text and elem blocks block with dynamic content and subblock 1`] = `
"class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      let ref1 = this.el.firstChild.nextSibling.firstChild;
      this.refs = [ref0, ref1];
      this.data = [];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentData = this.data;
      const nextData = data.data;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); } }
      else if (child0) { mountBefore(child0, refs[0]) }
      if (nextData[0] !== currentData[0]) { refs[1].textContent = nextData[0]}
      this.data = nextData;
      this.children = children;
    }
  }"
`;

exports[`text and elem blocks block with dynamic content and subblock 2`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks block with subblock 1`] = `
"class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      this.refs = [ref0];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); } }
      else if (child0) { mountBefore(child0, refs[0]) }
      this.children = children;
    }
  }"
`;

exports[`text and elem blocks block with subblock 2`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks block with subblock with dynamic content 1`] = `
"class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      this.refs = [ref0];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); } }
      else if (child0) { mountBefore(child0, refs[0]) }
      this.children = children;
    }
  }"
`;

exports[`text and elem blocks block with subblock with dynamic content 2`] = `
"class Builder {
    constructor(data) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild;
      this.refs = [ref0];
      this.data = [];
      this.update(data);
    }
    update(data) {
      const refs = this.refs;
      const currentData = this.data;
      const nextData = data.data;
      if (nextData[0] !== currentData[0]) { refs[0].textContent = nextData[0]}
      this.data = nextData;
    }
  }"
`;

exports[`text and elem blocks block with subblock with siblings 1`] = `
"class Builder {
    constructor(data, children) {
      this.el = template.cloneNode(true);
      let ref0 = this.el.firstChild.nextSibling;
      this.refs = [ref0];
      this.children = [];
      this.update(data, children);
    }
    update(data, children) {
      const refs = this.refs;
      const currentChildren = this.children;
      let currentChild0 = currentChildren[0], child0 = children[0];
      if (currentChild0) { if (child0) { patch(currentChild0, child0); } else { remove(currentChild0); } }
      else if (child0) { mountBefore(child0, refs[0]) }
      this.children = children;
    }
  }"
`;

exports[`text and elem blocks block with subblock with siblings 2`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks elem block can be removed 1`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks simple block with multiple roots 1`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks simple block with multiple roots 2`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;

exports[`text and elem blocks simple elem block 1`] = `
"class Builder {
    constructor() {
      this.el = template.cloneNode(true);
    }
    update() {
    }
  }"
`;
